### ISSUE TITLE
Workers or master may not start when other tasks exit in async distributed training

### ISSUE THREAD

**Issue by tobegit3hub** [tensorflow/tensorflow#21629]
### System information

== cat /etc/issue ===============================================
Darwin mbp-2.local 17.5.0 Darwin Kernel Version 17.5.0: Fri Apr 13 19:32:32 PDT 2018; root:xnu-4570.51.2~1/RELEASE_X86_64 x86_64
Mac OS X 10.13.4

== are we in docker =============================================
No

== compiler =====================================================
Apple LLVM version 9.1.0 (clang-902.0.39.1)
Target: x86_64-apple-darwin17.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

== uname -a =====================================================
Darwin mbp-2.local 17.5.0 Darwin Kernel Version 17.5.0: Fri Apr 13 19:32:32 PDT 2018; root:xnu-4570.51.2~1/RELEASE_X86_64 x86_64

== check pips ===================================================
numpy                              1.14.2
protobuf                           3.5.2.post1
tensorflow                         1.8.0
tensorflow-hub                     0.1.0
tensorflow-model-analysis          0.6.0
tensorflow-serving-api             1.8.0
tensorflow-tensorboard             1.5.0
tensorflow-transform               0.6.0
tensorflowjs                       0.1.0
tensorflowonspark                  1.0.0

== check for virtualenv =========================================
False

== tensorflow import ============================================
tf.VERSION = 1.8.0
tf.GIT_VERSION = v1.8.0-0-g93bc2e2072
tf.COMPILER_VERSION = v1.8.0-0-g93bc2e2072
Sanity check: array([1], dtype=int32)
/usr/local/lib/python2.7/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters

== env ==========================================================
LD_LIBRARY_PATH is unset
DYLD_LIBRARY_PATH is unset

== nvidia-smi ===================================================
tf_env_collect.sh: line 106: nvidia-smi: command not found

== cuda libs  ===================================================

== cat /etc/issue ===============================================
Darwin mbp-2.local 17.5.0 Darwin Kernel Version 17.5.0: Fri Apr 13 19:32:32 PDT 2018; root:xnu-4570.51.2~1/RELEASE_X86_64 x86_64
Mac OS X 10.13.4

== are we in docker =============================================
No

== compiler =====================================================
Apple LLVM version 9.1.0 (clang-902.0.39.1)
Target: x86_64-apple-darwin17.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

== uname -a =====================================================
Darwin mbp-2.local 17.5.0 Darwin Kernel Version 17.5.0: Fri Apr 13 19:32:32 PDT 2018; root:xnu-4570.51.2~1/RELEASE_X86_64 x86_64

== check pips ===================================================
numpy                              1.14.2
protobuf                           3.5.2.post1
tensorflow                         1.8.0
tensorflow-hub                     0.1.0
tensorflow-model-analysis          0.6.0
tensorflow-serving-api             1.8.0
tensorflow-tensorboard             1.5.0
tensorflow-transform               0.6.0
tensorflowjs                       0.1.0
tensorflowonspark                  1.0.0

== check for virtualenv =========================================
False

== tensorflow import ============================================
tf.VERSION = 1.8.0
tf.GIT_VERSION = v1.8.0-0-g93bc2e2072
tf.COMPILER_VERSION = v1.8.0-0-g93bc2e2072
Sanity check: array([1], dtype=int32)
/usr/local/lib/python2.7/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters

== env ==========================================================
LD_LIBRARY_PATH is unset
DYLD_LIBRARY_PATH is unset

== nvidia-smi ===================================================
./tf_env_collect.sh: line 106: nvidia-smi: command not found

== cuda libs  ===================================================

### Describe the problem

We have implement the async distributed TensorFlow training and find some issues. Once all the master(the worker with index 0), ps and workers start, the master will start training. And sometime(not always) one of the workers will halt and have no chance to end the progress.

Here is the [test distributed script](https://github.com/tobegit3hub/distributed_tensorflow/blob/master/auto_stop_ps/task.py) .

Here is the log of master which is normal and successfully finished.

<img width="1437" alt="screen shot 2018-08-15 at 5 07 00 pm" src="https://user-images.githubusercontent.com/2715000/44141773-bbfdb51e-a0b0-11e8-922a-c16e15d76637.png">

Here is the log of ps which is normal and join for all workers finished.

<img width="1439" alt="screen shot 2018-08-15 at 5 06 48 pm" src="https://user-images.githubusercontent.com/2715000/44141836-dda5391c-a0b0-11e8-82b0-a1390182454b.png">

Here is the log of worker which is abnormal. It failed to run since master has been finished. If the epoch is larger and master need more time to run, this worker may run and finish as well.

<img width="1439" alt="screen shot 2018-08-15 at 5 06 32 pm" src="https://user-images.githubusercontent.com/2715000/44141884-097955dc-a0b1-11e8-8f56-0e68d1478744.png">

### Source code / logs

Above.

**Comment by tobegit3hub** [comment]
Moreover, we have test in multi-node with unstable network. The worker may finish before master starting which results in master could never exit and export the model.

I don't think it's what we expect. Is there any way to guarantee the master will finish at any situation?

**Comment by cy89** [comment]
@mrry this seems like a "distributed training" issue. Can you please comment?

**Comment by mrry** [comment]
I can see two possibilities here:
* If one task depends on another then you need some mechanism to ensure that all tasks eventually come up at the same time, and retry in the event of failure. A cluster manager like Kubernetes or Slurm would be helpful here.
* If a task can run independently, then try setting `ConfigProto.device_filters` to contain only the device name prefixes for the devices it needs (e.g. the filters for the master task could be `["/job:master", "/job:ps"]` if the master task does not need to use the worker tasks).

**Comment by mattkindy-praetorian** [comment]
I've run into the same issue. As suggested by @mrry , I was able to resolve this by using device filters.

For example,
```
def _get_session_config_from_env_var():
  """Returns a tf.ConfigProto instance that has appropriate device_filters set.
  """

  tf_config = json.loads(os.environ.get('TF_CONFIG', '{}'))

  if (tf_config and 'task' in tf_config and 'type' in tf_config['task'] and
      'index' in tf_config['task']):
    # Master should only communicate with itself and ps
    if tf_config['task']['type'] == 'master':
      return tf.ConfigProto(device_filters=['/job:ps', '/job:master'])
    # Worker should only communicate with itself and ps
    elif tf_config['task']['type'] == 'worker':
      return tf.ConfigProto(device_filters=[
          '/job:ps',
          '/job:worker/task:%d' % tf_config['task']['index']
      ])
  return None
```
(originally from https://github.com/GoogleCloudPlatform/cloudml-samples/blob/master/census/estimator/trainer/task.py)

Outside of an estimator, you can use a `MonitoredTrainingSession`:
```
session = tf.train.MonitoredTrainingSession(config=_get_session_config_from_env_var(), ...)
``` 

With an estimator, instead, this just involves using
```
config = tf.estimator.RunConfig(session_config=_get_session_config_from_env_var())
```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by harshini-gadige** [comment]
@tobegit3hub Hi, please try using ConfigProto.device_filters and see if you are able to resolve this issue.

Note : Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

**Comment by tobegit3hub** [comment]
Sorry for the late response.

The `ConfigProto.device_filters` works work me as well. Thanks @mrry and @mattkindy-praetorian 😃
